src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -50,24 +51,28 @@ def run_python_test(transitive_hydrated_target, pytest):
       for py_req in maybe_python_req_lib.adaptor.requirements:
         all_requirements.append(str(py_req.requirement))
 
-  # TODO: This should be configurable, both with interpreter constraints, and for remote execution.
   # TODO(#7061): This str() can be removed after we drop py2!
   python_binary = text_type(sys.executable)
 
+  # TODO(EA): go from HydratedTarget to PythonTarget
```

Comment:
Ideas on how to do this? Danny mentioned using a `target_adaptor`, but I can't grok from the ones I see how this would work.

The goal is to coerce the `HydratedTarget`s into something with `.compatibility`, so that `python_setup.compatibility_or_constraints()` will work.

cc @stuhood @cosmicexplorer @illicitonion

Reply:
See just above: `for py_req in maybe_python_req_lib.adaptor.requirements`. The kwargs of a target constructor are made available via getattr:
https://github.com/pantsbuild/pants/blob/f30c612f7b9c70e0b1f4cf234d8c9155a8b27508/src/python/pants/engine/struct.py#L251-L253

So you can:
```python
if hasattr(ht.adaptor, 'compatibility'):
  ht.adaptor.compatibility  # Use it
```

Reply:
@Eric-Arellano : The v2 target API is still nascent, so it's not 100% clear what we want to do in these cases. One approach would be to add a second form of the method that takes a `TargetAdaptor` subclass... another might be to convert the existing method into something that takes Requirements rather than Targets.

Reply:
See https://github.com/pantsbuild/pants/pull/7680#issuecomment-490521656.

> another might be to convert the existing method into something that takes Requirements rather than Targets.

I don't want to modify `PythonSetup` at all if possible, which is option #4 in the comment. Instead, I would prefer to help with the V2 target API so that tasks like this become much easier to do for future development.

Reply:
I found a way around this that's acceptable. We decided to punt on #7680.